### DISCUSSION TITLE
Таблица-регистр или выборка из всех операций?

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#50] [Как это сделать?] [answered]
Здравствуйте!
Начал писать небольшой свой блок управленческого учета. В первую очередь стоит задача фиксации движения денежных средств.
Как правильнее это реализовать с точки зрения скорости и т.п.?
Изначально была идея создать некую таблицу - регистр, в которой будут висеть остатки по всем счетам и кассам. 
А изменения чтобы в ней проводились одновременно  с добавлением записи в таблице операций прихода-расхода.
Однако увидел, что в демо-сервере это реализовано по другому - в таблице движения по складу постоянно подсчитывается остаток ,и это значение потом уже подставляется через выборку с сортировкой.
Как правильнее это реализовать, какие подводные камни могут быть в первом моем описанном варианте?

**Comment by juramebel** [comment]
Понял, что в общем Ваш вариант примерно то же самое)
Но дает возможность потом использовать дополнительно статистику по изменению денег на конкретную дату. 
Поскольку вычисление поля в таблице товары (по сути тот же регистр) происходит кодом в его поле. 
Взял за основу ваш вариант.
Только добавил пересчет таблицы - регистра по добавлению строки в таблице операций.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Ага, видно, что вы уже запилили что-то работающее, из воздуха такой вопрос появиться не может 😄 

В чем проблема расчета остатка — в том, что нам надо взять список операций и его просуммировать. Если у нас будет 100 млн строк — это будет очень больно. Представляя количество операций у вас я предположу, что вы не столкнетесь с такой проблемой даже если сделаете простое суммирование от самой первой операции!

> Я пару дней назад нашел такую ошибку в одном из эксплуатирующихся проектов — брались все заказы а не только находящиеся в работе. Там за 4 года набежало 300К строк — замедление расчета составило 0.7 сек.

### Вариант 1 — расчет в момент операции

Тк остаток по кассе/счету/складу обычно используется достаточно часто — идеологически правильно иметь такую архитектуру, что бы скорость его расчета не замедлялась со временем эксплуатации. Поэтому мы почти везде используем расчет текущего остатка прям в строке операции на основе предыдущего значения:

```
=: #summ_oper + $prev

prev: select(table: $#ntn; field: 'remainder'; where: 'type' = #type; order: 'id' desc)

// этот код годится только вместе с параметром Исполнять только при добавлении!

```

Аналогичным решением является не `recalculate` строчку в справочнике (складе например), которая кодом берет значение из таблицы с операциями, а сразу делать туда `set` (он умеет относительные изменение `field: 'field' + #value`). Те вы рассчитываете остаток кодом выше и в этом же поле делаете код-действия по добавлению с `set`. У меня 50/50 в разных решениях.

### Вариант 2 — агломерат

Проблема описанных выше подходов (она же их преимущества) — иммутабельность данных, те невозможность провести изменение задним числом или исправлением. Поэтому все в строчке должно быть заблокировано на изменение — разрешено только добавление. Если произошла ошибка, то проводится строка с корректировкой.

Это не всегда подходит — когда не подходит разумно иметь отдельную таблицу с остатками **за период!**. Например по Кассе за месяц или за год (что бы ограничить количество строк попадающих в суммирование). Будет **простая таблица**:

| Счет | Период | Остаток |
|-|-|-|
| Касса 1 | Январь | Остаток |
| Касса 2 | Январь | Остаток |
| Касса 1 | Февраль | Остаток |
| Касса 2 | Февоаль | Остаток |

И в таком виде вы рекалкулэйтите нужную вам строку по счету и месяцу периода.

Месяц обычно делается по первой дате. Например `Январь = 2022-01-01`. Показывается как месяц через Формат отображения даты, а для сравнения получается таким способом:

```
=: dateformat(date: #date; format: "Y-m-01")
```

### Человекоподобная таблица с текущими остатками

Это уже надстройка над первым или вторым вариантом — обычно делается временная таблица. Используется только для показа пользователю, коды в нее не ходят — они идут либо в таблицу операций, либо в таблицу агломерата.
